### ISSUE TITLE
Passing C# objects with Blazor JS interop

### ISSUE THREAD

**Issue by pm100** [dotnet/AspNetCore.Docs#24646]
if I have

> public class Foo{
>   public String Name
> ...
> }

explain how to pass an instance of this to JS interop

 - pure fields do not work (they need settgers and getters)
 - the member names are changed to match JS naming conventions ie Name => name etc. (a big surprise)

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 2b7ea38a-649b-303d-bebf-c701e216ba25
* Version Independent ID: 9e197a8b-2616-8363-0a15-38cac802e82e
* Content: [Call JavaScript functions from .NET methods in ASP.NET Core Blazor](https://docs.microsoft.com/en-us/aspnet/core/blazor/javascript-interoperability/call-javascript-from-dotnet?view=aspnetcore-6.0#catch-javascript-exceptions)
* Content Source: [aspnetcore/blazor/javascript-interoperability/call-javascript-from-dotnet.md](https://github.com/dotnet/AspNetCore.Docs/blob/main/aspnetcore/blazor/javascript-interoperability/call-javascript-from-dotnet.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-blazor**
* GitHub Login: @guardrex
* Microsoft Alias: **riande**

**Comment by guardrex** [comment]
Hello @pm100 ... These are **_JSON-serializable_** `Object[]`s (.NET side) arguments. Therefore, I don't think it should be a surprise that serialization affects the name casing. Anyway ... on the broader point: They're being passed back and forth between what are essentially two runtimes. There will be limitations ... and probably some *gotchas* 😈, too 😄.

**_Yes!_** I think guidance should be added. I'll have to consider further 🤔 if an *Overview* topic section can cover both topics of the node or if each of the two topics should have a dedicated section. I'll also confer with the product unit on what to cover and where. Thanks for the feedback.

I'll get to this as soon as I can, but there could be a delay of several weeks. We have some priority work going on in the near-term that takes precedence at the moment. Don't worry tho ... this won't get lost. It's on the tracking project for Blazor docs.

**Comment by pm100** [comment]
Well it depends on whose serializer you are using. JSON.Net , the standard c# community serializer for the last 10 years, does not change names. So it surprised me.

At least include an example

**Comment by guardrex** [comment]
Yeah ... we can do better than that. I think this needs proper guidance and at least one example both ways, especially including any PU feedback for devs getting tripped up in known 😈 that either aren't supported.

I'll get on this as soon as I can. It just might take several weeks to reach it.

If you need product support in the meantime, you can hit up the community at the usual spots ...

* [Stack Overflow (tag: `blazor`)](https://stackoverflow.com/questions/tagged/blazor)
* [ASP.NET Core Slack Team](http://tattoocoder.com/aspnet-slack-sign-up/)
* [Blazor Gitter](https://gitter.im/aspnet/Blazor)

If you hit what looks like a bonafide bug/strange behavior or wish to ask the product unit a direct question, open an issue on their repo at ...

https://github.com/dotnet/aspnetcore/issues

Before asking tho, search all of the Blazor issues (`area-blazor` label), both open **_and closed_**. You might not be the first one with questions/concerns about it. They may even have an issue about improving aspects of the serialization. If you find some issues there potentially relevant to doc updates, you can cross-link them in a comment here. However, I will perform such a search before I even speak with them about this. You don't really need to go to any trouble on it.

If you open an issue on their repo, please include a ...

```
cc: @guardrex https://github.com/dotnet/AspNetCore.Docs/issues/24646
```

... at the bottom of your opening comment so that I can follow along and have it cross-linked to this for reference material.

**Comment by guardrex** [comment]
I'll be collecting some research notes, including general points/discussion not directly related to this issue's ask ...

* It uses `System.Text.Json`, which I don't think we say/cross-link (*yet*).

* > stay away from marshalling large and complex objects back and forth ... If you insist on going this route, you can serialize your payload as a byte array (UInt8Array) and use that when performing the interop call, we'll marshal it efficiently and you'll get full control over the serialization and deserialization process.

  **_SEE_** :point_right: https://github.com/dotnet/aspnetcore/issues/37312

* > `DateOnly`/`TimeOnly` require wrapping.

   **_SEE_** :point_right: 

  * https://github.com/dotnet/aspnetcore/issues/36631
  * https://github.com/dotnet/runtime/issues/53539

* > you can have an `IJSObjectReference` nested arbitrarily in your object graph, but you need your JS-side code to declare it

   **_SEE_** :point_right: https://github.com/dotnet/aspnetcore/issues/29981

* @pm100 ... I think this point answers one of your questions ...

  > The objects are being deserialized as JsonElement instances under the hood, so the behavior you see is by design. JsonElement is a C# representation of the json payload you sent and that has mixed casing.

   **_SEE_** :point_right: https://github.com/dotnet/aspnetcore/issues/18298

* > When serializing, with JS Interop, an object of a type which contains a property of an interface type, the resulting JSON for that property will contain only the properties defined on the interface, and not the ones defined on the actual implementing type.

  Looks resolved as of 6.0.

   **_SEE_** :point_right: https://github.com/dotnet/aspnetcore/issues/13474

* > support for customing serialization options per JS invocation

   **_SEE_** :point_right: 

  * https://github.com/dotnet/aspnetcore/issues/12685
  * https://github.com/dotnet/aspnetcore/issues/11954

* > When you do JS interop and pass JSON-serializable data, we JSON-serialize the data before it goes to {THING X}. As such, {THING X} only sees a string, not your original .NET object.

   **_SEE_** :point_right: https://github.com/dotnet/aspnetcore/issues/15644
